### PR TITLE
Add integration for tgstation-server

### PR DESCRIPTION
**Pull request by Cyberboss**
https://github.com/tgstation/tgstation-server being the production hosting tool most of SS13 uses.

I couldn't test this as tgui won't compile with `BUILD.bat`? But the DM code compiles fine. If someone could tell me how, the remaining TODO's are:

- [x] (External) Modify webpanel to allow for differing source/target pre-populate directories (for `config_example`)
- [x] Add linux scripts
- [x] Test TGS quick setup on Windows and Linux

You can safely ignore `__DEFINES/tgs.dm` and everything under `code/modules/tgs` for this review. It's taken directly from the latest DMAPI release on the TGS repo.

### CODE REVIEW

File: __DEFINES/tgs.config.dm [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+#define TGS_DEFINE_AND_SET_GLOBAL(Name, Value) var/global/##Name = ##Value
+#define TGS_READ_GLOBAL(Name) global.##Name
+#define TGS_WRITE_GLOBAL(Name, Value) global.##Name = ##Value
+#define TGS_WORLD_ANNOUNCE(message) to_chat(world, "<span class='boldannounce>[html_encode(##message)]</span>")
+#define TGS_INFO_LOG(message) log_world("TGS Info: [##message]")
+#define TGS_WARNING_LOG(message) log_world("TGS Warn: [##message]")
+#define TGS_ERROR_LOG(message) stack_trace("TGS Error: [##message]")
+#define TGS_NOTIFY_ADMINS(event) message_admins(##event)
+#define TGS_CLIENT_COUNT global.clients.len
+#define TGS_PROTECT_DATUM(Path) // This line is supposed to prevent VV editing/calling/fucking with the TGS glabal datum, but does /vg/ even have VV??
```
Comment (Cyberboss) [inline __DEFINES/tgs.config.dm:12]: This question could use answering

Comment (Kurfursten) [reply __DEFINES/tgs.config.dm:12]: Yes

Comment (Cyberboss) [reply __DEFINES/tgs.config.dm:12]: Is there a way to prevent VV on certain objects?

Comment (Kurfursten) [reply __DEFINES/tgs.config.dm:12]: Only administrators with +DEBUG are permitted to VV, we assume a high level of confidence in anyone given this flag. Then the actual usage on the menu is further broken down with other flags like +VAREDIT. We assume high levels of confidence in whoever has these flags, no restrictions.

Comment (Cyberboss) [reply __DEFINES/tgs.config.dm:12]: Easy enough to leave this blank then

**Comment by ShiftyRail** [comment]
Our existing version of TGS (I think?) is a bit old and rusty but it works. What's new with this one?

**Comment by d3athrow** [comment]
We don't use tgs, hina prompted cyberboss to add support for it because they likely don't know how to get the watchdog script working.

**Comment by Cyberboss** [comment]
Doesn't hurt to have regardless since it maintains itself :)

File: .github/workflows/update_tgs_dmapi.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      run: |
+        git branch -f tgs-dmapi-update
+        git checkout tgs-dmapi-update
+        git reset --hard master
+
+    - name: Apply DMAPI update
+      uses: tgstation/tgs-dmapi-updater@v2
+      with:
+        header-path: 'code/__DEFINES/tgs.dm'
+        library-path: 'code/modules/tgs'
```
Comment (Cyberboss) [inline .github/workflows/update_tgs_dmapi.yml]: Incorrect `header-path`

```suggestion
    - name: Apply DMAPI update
      uses: tgstation/tgs-dmapi-updater@v2
      with:
        header-path: '__DEFINES/tgs.dm'
        library-path: 'code/modules/tgs'
```

**Review by Cyberboss — COMMENTED** [review]
So, like the PR body says, how does one build /vg/ these days? `Ctrl+B` in VSC and `BUILD.cmd` don't seem to work

**Comment by d3athrow** [comment]
> So, like the PR body says, how does one build /vg/ these days? Ctrl+B in VSC and BUILD.cmd don't seem to work

Building in vsc works for me, it calls tools/build/build.bat

**Comment by DeityLink** [comment]
Alternatively just build in DM 😌

**Comment by Cyberboss** [comment]
> > So, like the PR body says, how does one build /vg/ these days? Ctrl+B in VSC and BUILD.cmd don't seem to work
> 
> Building in vsc works for me, it calls tools/build/build.bat

Isn't tgui part of your normal build?
This is what running `git clean -fxd; ./build.cmd` gets me:
```
S:\workspace\vgstation13 [tgs-support ≡]> git clean -fxd; .\BUILD.cmd
Removing tgui/.pnp.cjs
Removing tgui/.pnp.loader.mjs
Removing tgui/.yarn/cache/
Removing tgui/.yarn/install-state.gz
Removing tgui/.yarn/install-target
Removing tgui/.yarn/unplugged/
Removing vgstation13.dmb
Removing vgstation13.rsc
Using system-wide Node v20.7.0
:: Juke Build version 0.8.1
=> Starting 'dm'
=> Starting 'yarn'
:: Using defines: CBT
DM compiler version 515.1623
loading vgstation13.m.dme
➤ YN0000: ┌ Resolution step
➤ YN0000: └ Completed
➤ YN0000: ┌ Fetch step
➤ YN0013: │  - one package was already cached, 925 had to be fetched
➤ YN0000: └ Completed in 1s 38ms
➤ YN0000: ┌ Link step
loading interface/web/info.dms
loading interface/web/interface.dms
loading interface/skin.dmf
➤ YN0000: │ ESM support for PnP uses the experimental loader API and is therefore experimental
➤ YN0000: └ Completed in 4s 2ms
➤ YN0000: Done with warnings in 5s 197ms
=> Finished 'yarn' in 5.489s
=> Starting 'tgui'

node:internal/process/esm_loader:48
      internalBinding('errors').triggerUncaughtException(
                                ^
Error [ERR_UNSUPPORTED_ESM_URL_SCHEME]: Only URLs with a scheme in: file, data, and node are supported by the default ESM loader. On Windows, absolute paths must be valid file:// URLs. Received protocol 's:'
    at new NodeError (node:internal/errors:405:5)
    at throwIfUnsupportedURLScheme (node:internal/modules/esm/load:216:11)
    at defaultLoad (node:internal/modules/esm/load:118:3)
    at nextLoad (node:internal/modules/esm/hooks:832:28)
    at load$1 (file:///S:/workspace/vgstation13/tgui/.pnp.loader.mjs:1451:12)
    at nextLoad (node:internal/modules/esm/hooks:832:28)
    at Hooks.load (node:internal/modules/esm/hooks:415:26)
    at handleMessage (node:internal/modules/esm/worker:168:24)
    at checkForMessages (node:internal/modules/esm/worker:117:28)
    at process.<anonymous> (node:internal/modules/esm/worker:136:5) {
  code: 'ERR_UNSUPPORTED_ESM_URL_SCHEME'
}

Node.js v20.7.0
=> Target 'tgui' failed in 0.48s, exit code: 1
loading maps/tgstation.dmm
saving vgstation13.m.dmb (DEBUG mode)
vgstation13.m.dmb - 0 errors, 0 warnings (12/24/23 8:32 pm)
Total time: 0:13
=> Finished 'dm' in 13.818s
=> Target 'build' failed
=> Please inspect the error and close the window.
>
```

**Comment by Cyberboss** [comment]
See #35662

**Comment by Cyberboss** [comment]
Confirmed working (minus the separate node issue I logged)

**Comment by wotisjanitor** [comment]
we are not /tg/

**Comment by Cyberboss** [comment]
tgstation-server has basically nothing to do with /tg/

**Comment by d3athrow** [comment]
What is the significance of all those tgs_api datum procs returning tgs_unimplemented?

**Comment by Cyberboss** [comment]
> What is the significance of all those tgs_api datum procs returning tgs_unimplemented?

This provides support for TGS versions 3-6 (Interop versions 3-5). Not every feature is implemented in the previous versions. 

If you're asking about the base `/datum/tgs_api` definition, those are just stubs because DM doesn't have interfaces.

**Comment by Kurfursten** [comment]
The call on whether to use this should be entirely up to @d3athrow

**Comment by d3athrow** [comment]
I have no intentions of using it but no reason not to merge the pr

**Pull request merged**